runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeNotificationHandler.java

```diff
@@ -0,0 +1,116 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.grandexchange;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.LinkedList;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import net.runelite.api.GrandExchangeOffer;
+import net.runelite.api.GrandExchangeOfferState;
+import net.runelite.api.ItemComposition;
+
+@Singleton
+public class GrandExchangeNotificationHandler
+{
+	@Inject
+	private GrandExchangeConfig config;
+
+	private final LinkedList<GrandExchangeNotification> notificationQueue = new LinkedList<>();
+
+	private Instant lastNotificationSent = Instant.now();
+
+	public void addNotification(int slot, ItemComposition offerItem, GrandExchangeOffer newOffer)
+	{
+		// Get offer data
+		int itemId = offerItem.getId();
+		int quantitySold = newOffer.getQuantitySold();
+		int totalQuantity = newOffer.getTotalQuantity();
+		String itemName = offerItem.getName();
+
+		// Ignore empty/cancelled offers, or those with invalid data
+		if (newOffer.getState() == GrandExchangeOfferState.EMPTY ||
+			newOffer.getState() == GrandExchangeOfferState.CANCELLED_BUY ||
+			newOffer.getState() == GrandExchangeOfferState.CANCELLED_SELL ||
+			quantitySold == 0 || itemId == 0 || itemId == -1)
+		{
+			return;
+		}
+
+		// Inspect the queue for an existing notification from the same offer slot
+		GrandExchangeNotification notification = null;
+		for (GrandExchangeNotification existingNotification : notificationQueue)
+		{
+			if (existingNotification.getSlot() == slot)
+			{
+				notification = existingNotification;
+				break;
+			}
+		}
+
+		// See if we need to update the notification that's already in the queue (IE, if it goes from
+		// a partially sold offer to a fully sold offer within on set of events)
+		if (notification != null)
```

Comment:
It would be cleaner if the notifications were immutable and you simply removed/replaced it in the list.